docs/en/advanced_setup.md

```diff
@@ -275,3 +275,75 @@ FluentState::singleton()->withState(function (FluentState $state) {
 
 This is very important as global state is reverted back after the callback is executed so it's safe to be used.
 Unit tests benefit mostly from this as this makes sure that there are no dependencies between unit tests as the global state is always changed only locally in one test.
+
+## Top page reference feature (optional performance improvement)
+
+In some cases your project setup may have deeply nested blocks, for example:
+
+```
+Page
+  ElementalArea
+    RowBlock (represents grid row on frontend)
+     ElementalArea
+       AccordionBlock (block which can contain other content blocks)
+         ElementalArea
+           ContentBlock
+```
+
+It's quite common to use top page lookups from block context, i.e. a block is querying data from the page that the block belongs to.
+
+Most common cases are:
+
+* `CMS fields` - block level conditional logic depends on page data
+* `templates` - block level render logic depends on page data
+
+This module uses some in-memory caching but this isn't good enough for such deeply nested data structures by default.
+
+In such cases it is recommended to turn on this feature which stores the top page reference on individual blocks and elemental areas.
+This speeds up data lookup significantly.
+
+Please note that this feature only works with project setups which don't allow `block sharing`, i.e. `one block can only belong to a single page`.
+
+To turn the feature on simply apply following extensions, like this:
+
+```
+DNADesign\Elemental\Models\BaseElement:
+  extensions:
+    topPage: DNADesign\Elemental\TopPage\DataExtension
+
+DNADesign\Elemental\Models\ElementalArea:
+  extensions:
+    topPage: DNADesign\Elemental\TopPage\DataExtension
+
+Page:
+  extensions:
+    topPage: DNADesign\Elemental\TopPage\SiteTreeExtension
+```
+
+If your project setup uses Fluent module it is recommended to use following configuration instead:
+
+```
+DNADesign\Elemental\Models\BaseElement:
+  extensions:
+    topPage: DNADesign\Elemental\TopPage\FluentExtension
```

Comment:
Suggestion: Do we need to use the associative array syntax? This will mess with almost _every_ extension YML config I've seen which use indexed arrays for their extension syntax. If developers try to mix the two, the Symfony YML parser will barf. Also note Silverstripe's order of precedence when it comes to applying extensions. AFAIK, if developers attempt to declare >1 extension on a given class, only the latter will take and the former are clobbered.